### PR TITLE
 samples: Bluetooth: peripheral_hr and central_hr extended advertising support 

### PR DESCRIPTION
**Pull request by cvinayak**
Update peripheral_hr sample to support using extended
advertising and hence be long range compatible on SoCs
that support Bluetooth Low Energy Coded PHY.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: samples/bluetooth/central_hr/sample.yaml [hunk 1]
```diff
@@ -5,3 +5,15 @@ tests:
     arch_allow: x86
     harness: bluetooth
     tags: bluetooth
+  sample.bluetooth.central_hr.bt_ll_sw_split.phy_coded:
+    harness: bluetooth
+    platform_allow:
+      - nrf52_bsim
+      - nrf52dk/nrf52832
+      - nrf52840dk/nrf52840
```
Comment (aescolar) [inline samples/bluetooth/central_hr/sample.yaml:13]: nrf5340bsim_nrf5340_cpuapp too then :)

File: samples/bluetooth/peripheral_hr/overlay-phy_coded.conf [hunk 2]
```diff
@@ -0,0 +1,5 @@
+CONFIG_BT_AUTO_PHY_UPDATE=n
+CONFIG_BT_EXT_ADV=y
+
+CONFIG_BT_CTLR_ADV_DATA_LEN_MAX=191
```
Comment (Thalley) [inline samples/bluetooth/peripheral_hr/overlay-phy_coded.conf:4]: Seems like we set `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX=191` every time we do `CONFIG_BT_EXT_ADV=y`. Should we modify the default value of `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX`?

File: samples/bluetooth/peripheral_hr/src/main.c [hunk 3]
```diff
@@ -32,18 +33,28 @@ static const struct bt_data ad[] = {
 		      BT_UUID_16_ENCODE(BT_UUID_DIS_VAL))
 };
 
+/* Use atomic variable, 2 bits for connection and disconnection state */
+static ATOMIC_DEFINE(state, 2U);
+
+#define STATE_CONNECTED    1U
+#define STATE_DISCONNECTED 2U
```
Comment (Thalley) [inline samples/bluetooth/peripheral_hr/src/main.c:35]: What is the advantage of this over e.g. a `volatile bool`?

File: samples/bluetooth/peripheral_hr/src/main.c [hunk 4]
```diff
@@ -136,14 +132,73 @@ int main(void)
...
+		.sid = 0U,
+		.secondary_max_skip = 0U,
+		.options = (BT_LE_ADV_OPT_EXT_ADV |
+			    BT_LE_ADV_OPT_CONNECTABLE |
+			    BT_LE_ADV_OPT_CODED |
+			    BT_LE_ADV_OPT_USE_NAME),
+		.interval_min = BT_GAP_ADV_FAST_INT_MIN_2,
+		.interval_max = BT_GAP_ADV_FAST_INT_MAX_2,
+		.peer = NULL,
+	};
```
Comment (Thalley) [inline samples/bluetooth/peripheral_hr/src/main.c:196]: To me it makes more sense to use the same parameters, only add the `BT_LE_ADV_OPT_EXT_ADV` for the extended parameters. 

That way we can also use the same restart functionality for both configs.

File: samples/bluetooth/peripheral_hr/src/main.c [hunk 5]
```diff
@@ -152,6 +207,25 @@ int main(void)
 
 		/* Battery level simulation */
 		bas_notify();
+
+		if (atomic_test_and_clear_bit(state, STATE_CONNECTED)) {
+			/* Connected callback executed */
+
+		} else if (atomic_test_and_clear_bit(state,
+						     STATE_DISCONNECTED)) {
```
Comment (Thalley) [inline samples/bluetooth/peripheral_hr/src/main.c]: This looks weird. 

When we are connected, state is set to `STATE_CONNECTED`, but we then clearing it here? So after one iteration, we are still connected, but `STATE_CONNECTED` is not set.

File: samples/bluetooth/peripheral_hr/src/main.c [hunk 6]
```diff
@@ -152,6 +207,25 @@ int main(void)
...
+#if defined(CONFIG_BT_EXT_ADV)
+			printk("Starting Extended Advertising "
+			       "(connectable and non-scannable)\n");
+			err = bt_le_ext_adv_start(adv,
+						  BT_LE_EXT_ADV_START_DEFAULT);
+			if (err) {
+				printk("Failed to start extended advertising "
+				       "set (err %d)\n", err);
+				return 0;
+			}
```
Comment (Thalley) [inline samples/bluetooth/peripheral_hr/src/main.c:278]: Please use the recommended line length of `100`

File: samples/bluetooth/peripheral_hr/prj.conf [hunk 7]
```diff
@@ -8,3 +8,5 @@ CONFIG_BT_BAS=y
 CONFIG_BT_HRS=y
 CONFIG_BT_DEVICE_NAME="Zephyr Heartrate Sensor"
 CONFIG_BT_DEVICE_APPEARANCE=833
+
+CONFIG_GPIO=y
```
Comment (Thalley) [inline samples/bluetooth/peripheral_hr/prj.conf:12]: Does enabling this for a board that does not have LED, increase memory usage without actually adding any new functionality?

File: samples/bluetooth/peripheral_hr/src/main.c [hunk 8]
```diff
@@ -122,10 +117,50 @@ static void hrs_notify(void)
...
+#define LED0_NODE DT_ALIAS(led0)
+
+#if DT_NODE_HAS_STATUS(LED0_NODE, okay)
+#include <zephyr/drivers/gpio.h>
+#define HAS_LED     1
+static const struct gpio_dt_spec led = GPIO_DT_SPEC_GET(LED0_NODE, gpios);
+#define BLINK_ONOFF K_MSEC(500)
+
+static struct k_work_delayable blink_work;
+static bool                  led_is_on;
```
Comment (Thalley) [inline samples/bluetooth/peripheral_hr/src/main.c:130]: `led_is_on` seems to be modify by multiple threads. Should be `volatile` or have some sort of thread safety added

File: samples/bluetooth/central_hr/src/main.c [hunk 9]
```diff
@@ -168,15 +185,23 @@ static void start_scan(void)
...
 		.interval   = BT_GAP_SCAN_FAST_INTERVAL,
 		.window     = BT_GAP_SCAN_FAST_WINDOW,
 	};
 
 	err = bt_le_scan_start(&scan_param, device_found);
 	if (err) {
-		printk("Scanning failed to start (err %d)\n", err);
-		return;
+		printk("Scanning with Coded PHY support failed "
+		       "(err %d)\n", err);
```
Comment (Thalley) [inline samples/bluetooth/central_hr/src/main.c:200]: Please use recommended line length of `100`

File: samples/bluetooth/central_hr/src/main.c [hunk 10]
```diff
@@ -153,9 +167,12 @@ static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
...
 	       dev, type, ad->len, rssi);
 
-	/* We're only interested in connectable events */
+	/* We're only interested in legacy connectable events or
+	 * possible extended advertising that are connectable.
+	 */
 	if (type == BT_GAP_ADV_TYPE_ADV_IND ||
-	    type == BT_GAP_ADV_TYPE_ADV_DIRECT_IND) {
+	    type == BT_GAP_ADV_TYPE_ADV_DIRECT_IND ||
+	    type == BT_GAP_ADV_TYPE_EXT_ADV) {
```
Comment (Thalley) [inline samples/bluetooth/central_hr/src/main.c:179]: In the case that it is extended, we should filter additionally on `BT_GAP_ADV_PROP_CONNECTABLE`

File: samples/bluetooth/central_hr/boards/nrf5340bsim_nrf5340_cpuapp.conf [hunk 11]
```diff
@@ -0,0 +1 @@
+CONFIG_BT_BUF_ACL_RX_SIZE=255
```
Comment (Thalley) [inline samples/bluetooth/central_hr/boards/nrf5340bsim_nrf5340_cpuapp.conf:1]: Why do we need this change?

File: samples/bluetooth/central_hr/Kconfig.sysbuild [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+config NET_CORE_BOARD
+	string
+	default "nrf5340dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340dk"
+	default "nrf5340_audio_dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340_audio_dk"
+	default "nrf5340bsim_nrf5340_cpunet" if "$(BOARD)" = "nrf5340bsim_nrf5340_cpuapp"
+
+config NET_CORE_IMAGE_HCI_IPC
+	bool "HCI IPC image on network core"
+	default y
+	depends on NET_CORE_BOARD != ""
```
Comment (Thalley) [inline samples/bluetooth/central_hr/Kconfig.sysbuild:15]: This seems unrelated to testing. Suggest to move to the earlier commit that adds nRF5340 support for the samples.

File: tests/bsim/bluetooth/samples/central_hr_peripheral_hr/boards/nrf5340bsim_nrf5340_cpuapp.conf [hunk 13]
```diff
@@ -0,0 +1 @@
+CONFIG_BT_BUF_ACL_RX_SIZE=255
```
Comment (Thalley) [inline tests/bsim/bluetooth/samples/central_hr_peripheral_hr/boards/nrf5340bsim_nrf5340_cpuapp.conf:1]: Why do we need this both here and in the sample?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bsim/bluetooth/samples/compile.sh [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+mkdir -p ${WORK_DIR}
+
+source ${ZEPHYR_BASE}/tests/bsim/compile.source
+
+if [ "${BOARD}" == "nrf5340bsim_nrf5340_cpuapp" ]; then
+  app=samples/bluetooth/peripheral_hr \
+    sysbuild=1 \
+    conf_overlay=overlay-phy_coded.conf \
+    exe_name=bs_${BOARD}_${app}_prj_conf \
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply tests/bsim/bluetooth/samples/compile.sh]: @aescolar Now the name length is too long in the script file, should I reduce the name `central_hr_peripheral_hr` or do you have any suggestion so that the executable name does not execeed the 100 char column limit in the compliance?

[the remark this answers is not in the document]
Comment (aescolar) [reply tests/bsim/bluetooth/samples/compile.sh]: Oh well.. silly checks
If you want you can replace:
`Execute ./bs_${BOARD}_tests_bsim_bluetooth_samples_central_hr_peripheral_hr_prj_conf_overlay-phy_coded_conf \`
with
`Execute ./bs_${BOARD_TS}_$(guess_test_long_name)_prj_conf_overlay-phy_coded_conf \`
(Note that due to hwmv2 now we need to use BOARD_TS to ensure we don't have "/" in the name.

**Comment by aescolar** [comment]
@cvinayak how are we doing in this PR? Having some coverage for coded phy in CI would be very nice

**Comment by Thalley** [comment]
> @cvinayak how are we doing in this PR? Having some coverage for coded phy in CI would be very nice

I think these samples are pretty common to run for first-time users - It would be great to have them tested in BSIM in CI :)

**Comment by cvinayak** [comment]
> > @cvinayak how are we doing in this PR? Having some coverage for coded phy in CI would be very nice
> 
> I think these samples are pretty common to run for first-time users - It would be great to have them tested in BSIM in CI :)

Using these samples (Extended Scanning) on target has reveal an assertion and its analysis has caused the delay, hope to rework on this PR as soon as possible.

File: samples/bluetooth/peripheral_hr/prj.conf (hunk 7 above)
Comment (cvinayak) [reply samples/bluetooth/peripheral_hr/prj.conf]: Not able to test, as all my boards have LED. But agree that logically there could be more code added to boards without LED when GPIO module being included.

Removed explicit enable in the prj.conf, but depend on boards that already enable GPIO be default, like most nRF DK boards. In this case, LED  will be used from devicetree, when enabled by default.

File: samples/bluetooth/peripheral_hr/src/main.c (hunk 3 above)
Comment (cvinayak) [reply samples/bluetooth/peripheral_hr/src/main.c:35]: As callback and the main threads are at different priority, use of atomic flags is context safe compared to use of `volatile bool`

File: samples/bluetooth/peripheral_hr/src/main.c (hunk 8 above)
Comment (cvinayak) [reply samples/bluetooth/peripheral_hr/src/main.c:131]: The work is stopped (or not started) before the variable is modifed and hence it is safe.

File: samples/bluetooth/peripheral_hr/src/main.c (hunk 4 above)
Comment (cvinayak) [reply samples/bluetooth/peripheral_hr/src/main.c:198]: Will keep as-is as this sample shows the use of two different API parameters, use of public define versus local parameter variable.

File: samples/bluetooth/peripheral_hr/src/main.c (hunk 6 above)
Comment (cvinayak) [reply samples/bluetooth/peripheral_hr/src/main.c]: done.

File: tests/bsim/bluetooth/samples/central_hr_peripheral_hr/boards/nrf5340bsim_nrf5340_cpuapp.conf (hunk 13 above)
Comment (cvinayak) [reply tests/bsim/bluetooth/samples/central_hr_peripheral_hr/boards/nrf5340bsim_nrf5340_cpuapp.conf:1]: updated with a comment in the file

File: tests/bsim/compile.source [hunk 15]
```diff
@@ -19,6 +19,7 @@ function _compile(){
   local app_root="${app_root:-${ZEPHYR_BASE}}"
   local BOARD_ROOT="${BOARD_ROOT:-${ZEPHYR_BASE}}"
   local conf_file="${conf_file:-prj.conf}"
+  local extra_conf_file="${extra_conf_file:-""}"
```
Comment (aescolar) [inline tests/bsim/compile.source:22]: Nit: It would be nice to separate this file change in its own commit

**Review by aescolar — DISMISSED** [review]
Thanks @cvinayak a few changes are needed

File: samples/bluetooth/central_hr/Kconfig.sysbuild [hunk 16]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2023-2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+source "share/sysbuild/Kconfig"
+
+config NET_CORE_BOARD
+	string
+	default "nrf5340dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340dk"
+	default "nrf5340_audio_dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340_audio_dk"
+	default "nrf5340bsim_nrf5340_cpunet" if "$(BOARD)" = "nrf5340bsim_nrf5340_cpuapp"
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply samples/bluetooth/central_hr/Kconfig.sysbuild]: Done

File: samples/bluetooth/peripheral_hr/Kconfig.sysbuild [hunk 17]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2023-2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+source "share/sysbuild/Kconfig"
+
+config NET_CORE_BOARD
+	string
+	default "nrf5340dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340dk"
+	default "nrf5340_audio_dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340_audio_dk"
+	default "nrf5340bsim_nrf5340_cpunet" if "$(BOARD)" = "nrf5340bsim_nrf5340_cpuapp"
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply samples/bluetooth/peripheral_hr/Kconfig.sysbuild]: done

File: samples/bluetooth/central_hr/src/main.c (hunk 10 above)
Comment (Thalley) [reply samples/bluetooth/central_hr/src/main.c:178]: The comment above doesn't match the code. 

This will still call `bt_data_parse` on non-connectable extended advertising, right?

File: tests/bsim/bluetooth/samples/central_hr_peripheral_hr/Kconfig.sysbuild [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+# source the samples' Kconfig.sysbuild, if any
+source "${ZEPHYR_BASE}/samples/bluetooth/central_hr/Kconfig.sysbuild"
+
+config NATIVE_SIMULATOR_PRIMARY_MCU_INDEX
+	int
+	# Let's pass the test arguments to the application MCU test
+	# otherwise by default they would have gone to the net core.
+	default 0 if $(BOARD) = "nrf5340bsim_nrf5340_cpuapp"
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply tests/bsim/bluetooth/samples/central_hr_peripheral_hr/Kconfig.sysbuild]: done

File: tests/bsim/bluetooth/samples/central_hr_peripheral_hr/tests_scripts/central_hr_peripheral_hr_extended.sh [hunk 19]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+# Copyright 2023-2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+# Simple selfchecking test for the central_hr/peripheral_hr samples,
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply tests/bsim/bluetooth/samples/central_hr_peripheral_hr/tests_scripts/central_hr_peripheral_hr_extended.sh]: Removed

File: tests/bsim/bluetooth/samples/compile.sh [hunk 20]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env bash
+# Copyright 2023-2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+# Compile all the applications needed by the bsim tests in these subfolders
+
+#set -x #uncomment this line for debugging
+set -ue
+
+: "${BSIM_COMPONENTS_PATH:?BSIM_COMPONENTS_PATH must be defined}"
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply tests/bsim/bluetooth/samples/compile.sh]: Removed

File: samples/bluetooth/peripheral_hr/src/main.c [hunk 21]
```diff
@@ -126,24 +117,126 @@ static void hrs_notify(void)
...
+{
+	led_is_on = !led_is_on;
+	gpio_pin_set(led.port, led.pin, (int)led_is_on);
+
+	k_work_schedule(&blink_work, BLINK_ONOFF);
+}
+#endif
+#endif /* CONFIG_GPIO */
+
 int main(void)
```
Comment (Thalley) [inline samples/bluetooth/peripheral_hr/src/main.c:143]: `main` is now 150 lines of code. Suggest to move some of the code blocks into functions, so you have functions like

```
configure_gpio
start_advertising
start_led
stop_led
```

to make the flow of the `main` easier to follow.

File: tests/bsim/bluetooth/samples/central_hr_peripheral_hr/tests_scripts/central_hr_peripheral_hr.sh [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+# It relies on the bs_tests hooks to register a test timer callback, which after a deadline
+# will check how many packets has been received, and if over a threshold
+# it considers the test passed
+
+simulation_id="central_hr_peripheral_hr_test"
+verbosity_level=2
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+EXECUTE_TIMEOUT=200
```
Comment (Thalley) [inline tests/bsim/bluetooth/samples/central_hr_peripheral_hr/tests_scripts/central_hr_peripheral_hr.sh:15]: Why such a long timeout?

File: tests/bsim/bluetooth/samples/compile.sh [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+: "${ZEPHYR_BASE:?ZEPHYR_BASE must be set to point to the zephyr root\
+ directory}"
+
+WORK_DIR="${WORK_DIR:-${ZEPHYR_BASE}/bsim_out}"
+
+BOARD_ROOT="${BOARD_ROOT:-${ZEPHYR_BASE}}"
+
+mkdir -p ${WORK_DIR}
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply tests/bsim/bluetooth/samples/compile.sh]: Removed

File: tests/bsim/bluetooth/samples/compile.sh [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+
+source ${ZEPHYR_BASE}/tests/bsim/compile.source
+
+if [ "${BOARD_TS}" == "nrf5340bsim_nrf5340_cpuapp" ]; then
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply tests/bsim/bluetooth/samples/compile.sh]: Removed duplicates

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bsim/compile.source (hunk 15 above)
Comment (cvinayak) [reply tests/bsim/compile.source:22]: done

File: samples/bluetooth/central_hr/src/main.c (hunk 10 above)
Comment (cvinayak) [reply samples/bluetooth/central_hr/src/main.c:178]: yes, but that should be ok as the connection request is only sent to peripheral with HR service listed in its services AD type

Comment (Thalley) [reply samples/bluetooth/central_hr/src/main.c:178]: > yes, but that should be ok as the connection request is only sent to peripheral with HR service listed in its services AD type

A device can advertise HR without being connectable. I don't see any reason why we wouldn't add the connectable check so that we don't have a device that may attempt to connect to non-connectable advertising

File: tests/bsim/bluetooth/samples/central_hr_peripheral_hr/src/sample_test.c [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+
+#include "bs_types.h"
+#include "bs_tracing.h"
+#include "bs_utils.h"
+#include "time_machine.h"
+#include "bstests.h"
+
+#define WAIT_TIME 120 /* Seconds */
+
+#define PASS_THRESHOLD 100 /* packets */
```
Comment (Thalley) [inline tests/bsim/bluetooth/samples/central_hr_peripheral_hr/src/sample_test.c:16]:
```suggestion
#define PASS_THRESHOLD 5 /* packets */
```

I don't see any reason to wait for 100 notifications. The behavior should be verifiable with a few notifications

File: tests/bsim/bluetooth/samples/central_hr_peripheral_hr/src/sample_test.c [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+		bs_trace_exit("Done, disconnecting from simulation\n");
+	} else {
+		FAIL("FAILED (Did not pass after %i seconds)\n", WAIT_TIME);
+	}
+}
+
+static const struct bst_test_instance test_sample[] = {
+	{
+		.test_id = "central_hr_peripheral_hr",
+		.test_descr = "Test based on a sample. "
```
Comment (Thalley) [inline tests/bsim/bluetooth/samples/central_hr_peripheral_hr/src/sample_test.c:62]:
```suggestion
		.test_descr = "Test based on the peripheral and central HR samples. "
```

**Review by Thalley — DISMISSED** [review]
The code is OK, but I would heavily suggest to reduce the runtime of the tests

**Review by Thalley — DISMISSED** [review]

**Review by aescolar — DISMISSED** [review]

File: tests/bsim/bluetooth/samples/central_hr_peripheral_hr/src/sample_test.c (hunk 26 above)
Comment (cvinayak) [reply tests/bsim/bluetooth/samples/central_hr_peripheral_hr/src/sample_test.c]: missed this one... pushed again.

**Review by aescolar — APPROVED** [review]

**Review by Thalley — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]
